src/compiler/checker.ts

```diff
@@ -2587,6 +2532,15 @@ namespace ts {
             }
         }
 
+        // Aliases that resolve to const enums are not marked as referenced because they are not emitted,
+        // but their usage in value positions must be tracked to determine if the import can be type-only.
+        function markConstEnumAliasAsReferenced(symbol: Symbol) {
+            const links = getSymbolLinks(symbol);
+            if (!links.constEnumReferenced) {
+                links.constEnumReferenced = true;
+            }
+        }

```

Comment:
This was surprisingly the ugliest part of the changes. Aliases that resolve to const enums are explicitly excluded from being marked as `symbol.referenced` because that’s what the emit resolver uses to determine if the import can be elided, and since const enums are inlined, the alias is, in a sense, not referenced. Meanwhile, the `symbol.isReferenced` property, which is used to report unused variables, is incapable of tracking whether an enum was referenced as a type or as a value, because name resolution of dotted names like `Enum.Member` looks for the meaning `SymbolFlags.Namespace`, which overlaps with both `SymbolFlags.Type` and `SymbolFlags.Value`. The distinction is necessary for `--importsNotUsedAsValues=error`, unless we want to treat const enums in some special way under that flag.

Does this make sense, or did I miss an easier/better way to do this?

(I wonder if const enums should be marked as `referenced` during checking, and the logic to treat them as unreferenced be deferred to the emit resolver?)

Reply:
Proposal: Treat type only and normal imports the same in the checker, this way usage diagnostics are reported, but then in the transformer, always elide `import type` and `export type` statements in the TS transform (without needing to refer to the checker via the emit resolver).

Reply:
> this way usage diagnostics are reported

They already are.

> but then in the transformer, always elide `import type` and `export type` statements in the TS transform

This is already how it works, both in #35200 and here. What’s new here is that since we retain the value meaning of aliases, the emit resolver started returning `true` for `isValueAliasDeclaration` and `isReferencedAliasDeclaration` if those aliases are used in expressions (which is an error). The effect was kind of a weird emit. I wasn’t sure if it was worth bothering with since the code had errors, but @DanielRosenwasser suggested so:

![image](https://user-images.githubusercontent.com/3277153/72103494-2dc16900-32de-11ea-9d4d-20e9cccf0cec.png)

However, this is a little beside the point of my question here—I’m talking about the `symbol.referenced` of const enums, which has been a long-standing behavior before any of my type-only work.

Reply:
Right - my point is that if you ignore the emit resolver entirely when printing `import type| and `export type` and always the behavior of `symbol.referenced` no longer should matter, since you wouldn't be referencing it in the import/export type case. The references to those things, should they exist in value space, are fine to have the emit that expects the import be present, imo - the code's nonsensical anyway.

But anyways, re: the distinction - I'm pretty sure I remember at some point in the past Nathan had swapped `symbol.isReferenced` from a Boolean to `SymbolFags` encoding how it was referenced for a PR at some point. Doing that here could work well.

Reply:
If the nonsensical `a_1.A` emit reference is ok, then yes, I can make a couple simplifications that might even help the perf situation. But as far as the const enums go, from my top comment:

> the symbol.isReferenced property, which is used to report unused variables, is incapable of tracking whether an enum was referenced as a type or as a value, because name resolution of dotted names like `Enum.Member` looks for the meaning `SymbolFlags.Namespace`, which overlaps with both `SymbolFlags.Type` and `SymbolFlags.Value`. 

So if a const enum is referenced in a type position, it _appears_ as if it was also referenced in a value position, as far as `isReferenced` goes. I could probably try to tighten up that namespace lookup.

Reply:
> Doing that here could work well.

Oh, I may have misunderstood your suggestion. I thought you were suggesting to use `isReferenced`, but I think you may have been talking about changing `referenced` to flags.